File: day12/solution.py
# nice and concise solution inspired by this comment on reddit:
# https://www.reddit.com/r/adventofcode/comments/1hcdnk0/2024_day_12_solutions/m1nj6se/

type Pos = complex
type Direction = complex
type Region = frozenset[Pos]
type Edge = tuple[Pos, Direction]

DIRECTIONS = (1, -1, 1j, -1j)

def neighbors_of(p: Pos) -> list[Pos]:
    return [p + d for d in DIRECTIONS]

def regions_of(grid: dict[Pos, str]) -> set[Region]:
    regions = {pos: {pos} for pos in grid}
    for pos, this in grid.items():
        for npos in neighbors_of(pos):
            if grid.get(npos) == this:
                regions[pos] |= regions[npos]
                for p in regions[npos]:
                    regions[p] = regions[pos]
    return {frozenset(region) for region in regions.values()}

def edges_of(region: Region) -> set[Edge]:
    return {(p, d) for p in region for d in DIRECTIONS if (p + d) not in region}

def sides_of(region: Region) -> set[Edge]:
    edges = edges_of(region)
    return {(pos, d) for pos, d in edges if (pos + d * 1j, d) not in edges}

with open("input.txt") as f:
    lines = f.read().split()

grid = {
    x + y * 1j: ch
    for y, row in enumerate(lines)
    for x, ch in enumerate(row)
}
regions = regions_of(grid)

part1 = sum(len(region) * len(edges_of(region)) for region in regions)
print("Part 1:", part1)

part2 = sum(len(region) * len(sides_of(region)) for region in regions)
print("Part 2:", part2)

File: day12/solution.hs
import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!?))
import qualified Data.Set as Set
import Data.Set (Set)
import System.IO (readFile')

data Pos = P Int Int deriving (Show, Ord, Eq)
data Direction = North | East | South | West deriving (Eq, Ord, Enum, Bounded, Show)
type Grid a = Map Pos a
type Region = Set Pos
data Edge = Edge Pos Direction deriving (Show, Ord, Eq)

makeGrid :: String -> Grid Char
makeGrid s = Map.fromList
  [ (P x y, color)
  | (line, y) <- zip (lines s) [1..]
  , (color, x) <- zip line [1..]
  ]

neighbors :: Pos -> [Pos]
neighbors pos = [move d pos | d <- [minBound..maxBound]]

move :: Direction -> Pos -> Pos
move North (P x y) = P x (y-1)
move East  (P x y) = P (x+1) y
move South (P x y) = P x (y+1)
move West  (P x y) = P (x-1) y

rot90 :: Direction -> Direction
rot90 x = if x == maxBound then minBound else succ x

regionsOf :: Eq a => Grid a -> [Region]
regionsOf grid = go (Map.keys grid) mempty mempty
  where
    go :: [Pos] -> Set Pos -> [Region] -> [Region]
    go [] _ g = g
    go (p:ps) visited regions
      | Set.member p visited = go ps visited regions
      | otherwise = go ps (Set.union visited region) (region:regions)
      where region = floodfillFrom p mempty
    floodfillFrom :: Pos -> Set Pos -> Set Pos
    floodfillFrom p region
      | Set.member p region = region
      | otherwise = recur $ Set.insert p region
      where
        sameColor npos = (grid!?p) == (grid!?npos)
        recur = foldl (.) id $ floodfillFrom <$> filter sameColor (neighbors p)

edgesOf :: Region -> Set Edge
edgesOf region = Set.fromList
  [ Edge pos direction
  | pos <- Set.elems region
  , direction <- [minBound..maxBound]
  , Set.notMember (move direction pos) region
  ]

sidesOf :: Region -> Set Edge
sidesOf region = Set.filter ((`Set.notMember` edges) . adjacentEdge) edges
  where
    edges = edgesOf region
    adjacentEdge (Edge pos d) = Edge (move (rot90 d) pos) d

fenceCostPart1 :: Region -> Int
fenceCostPart1 region = length region * length (edgesOf region)

fenceCostPart2 :: Region -> Int
fenceCostPart2 region = length region * length (sidesOf region)

main :: IO ()
main = do
  input <- readFile' "input.txt"
  let grid = makeGrid input
  let regions = regionsOf grid
  let part1 = sum $ fenceCostPart1 <$> regions
  putStrLn $ "Part 1: " <> show part1
  let part2 = sum $ fenceCostPart2 <$> regions
  putStrLn $ "Part 2: " <> show part2
